src/typechecker.erl

```diff
@@ -4495,21 +4439,32 @@ handle_type_error(type_error, _) ->
 
 -spec describe_expr(gradualizer_type:abstract_expr()) -> string().
 describe_expr({atom, _, _})               -> "atom";
+describe_expr({bc, _, _, _})              -> "binary comprehension";
 describe_expr({bin, _, _})                -> "bit expression";
+describe_expr({block,_,_})                -> "block";
 describe_expr({char, _, _})               -> "character";
+describe_expr({call, _, _})               -> "function call";
+describe_expr({'catch', _, _})            -> "catch";
+describe_expr({'case', _, _, _})          -> "case";
 describe_expr({cons, _, _, _})            -> "list";
 describe_expr({float, _, _})              -> "float";
 describe_expr({'fun', _, _})              -> "fun";
 describe_expr({integer, _, _})            -> "integer";
+describe_expr({'if', _, _})               -> "if";
+describe_expr({lc, _, _, _})              -> "list comprehension";
 describe_expr({map, _, _})                -> "map";
-describe_expr({map, _, _, _} )            -> "map update";
+describe_expr({map, _, _, _})             -> "map update";
+describe_expr({match, _, _, _})           -> "match";
```

Comment:
Can you suggest what pattern to use to make this distinction? When does match show up and when does var show up?

Reply:
Perhaps we can give a more exact error message depending on whether they are free or bound variables? Another place it could occur is when the match expression itself is used where an expression of a different type is expected, e.g. `list_to_integer(X = 3.14)`. In this latter case, the match expression has the type of the RHS, so perhaps we can take the RHS out so the error says something like `The float 3.14 is expected to have the type list() but has type float()` for this example.

Anyhow, an expression on this form is called "match" or "match expression" in the Erlang manual, so I think it is right for the general case. Users might actually know it by this name. :-)

Reply:
Hmmm, it is indeed called match expression in the manual. I'm not a big fan of this terminology but I'll accept it as an erlangism.